paychmgr/simple.go

```diff
@@ -135,18 +151,97 @@ func (m *mergedFundsReq) onComplete(res *paychFundsRes) {
 }
 
 // sum is the sum of the amounts in all requests in the merge
-func (m *mergedFundsReq) sum() types.BigInt {
+func (m *mergedFundsReq) sum() (types.BigInt, types.BigInt) {
 	sum := types.NewInt(0)
+	avail := types.NewInt(0)
+
 	for _, r := range m.reqs {
 		if r.isActive() {
 			sum = types.BigAdd(sum, r.amt)
+			if !r.opts.Reserve {
+				avail = types.BigAdd(avail, r.amt)
+			}
+		}
+	}
+
+	return sum, avail
+}
+
+// completeAmount completes first non-reserving requests up to the available amount
+func (m *mergedFundsReq) completeAmount(avail types.BigInt, channelInfo *ChannelInfo) (*paychFundsRes, types.BigInt, types.BigInt) {
+	used, failed := types.NewInt(0), types.NewInt(0)
+	next := 0
+
+	// order: [offchain+reserve, !offchain+reserve, !offchain+!reserve]
+	for i, r := range m.reqs {
+		if !r.opts.Reserve {
+			// non-reserving request are put after reserving requests, so we are done here
+			break
+		}
+
+		if r.amt.GreaterThan(types.BigSub(avail, used)) {
+			// requests are sorted by amount ascending, so if we hit this, there aren't any more requests we can fill
+
+			if r.opts.OffChain {
+				// can't fill, so OffChain want an error
+				if r.isActive() {
+					failed = types.BigAdd(failed, r.amt)
+					r.onComplete(&paychFundsRes{
+						channel: *channelInfo.Channel,
+						err:     xerrors.Errorf("not enough funds available in the payment channel %s; add funds with 'lotus paych add-funds %s %s %s'", channelInfo.Channel, channelInfo.from(), channelInfo.to(), types.FIL(r.amt).Unitless()),
```

Comment:
Suggest simplifying this error message - the message only makes sense if people are calling the API from the command line